service-base/src/main/java/org/eclipse/hono/service/metric/MicrometerBasedMetrics.java

```diff
@@ -220,6 +233,20 @@ public final void reportTelemetry(
         }
     }
 
+    @Override
+    public void removeTelemetryMetricsForTenant(final MetricsTags.EndpointType type, final String tenantId) {
+        Objects.requireNonNull(type);
+        Objects.requireNonNull(tenantId);
+
+        if (type != MetricsTags.EndpointType.TELEMETRY && type != MetricsTags.EndpointType.EVENT) {
+            throw new IllegalArgumentException("invalid type, must be either telemetry or event");
+        }
+
+        final Tags tags = Tags.of(MetricsTags.getTenantTag(tenantId), type.asTag());
+        registry.find(METER_MESSAGES_PAYLOAD).tags(tags).meters().forEach(registry::remove);
```

Comment:
Are we sure that this code is thread safe?

Reply:
MeterRegistry#remove(Meter) does synchronization, but the timeout event, the cleanup and the creation of a new  entries are of course not synchronized together. I don't think that it would be a major problem if a tenant that has been inactive for e.g. one hour (default timeout) really happens to send a new message _during_ this cleanup and this single event is not reflected in the metrics (which are probably not 100% accurate anyways).

Reply:
I am not so worried about correct metrics, more about he facts that two thread could update the same, backing collection. Depending on the collection it may be that concurrent iterations and removing/adding can become a problem, and throw a CME. Which might even be the case in a single thread when you have multiple entries to remove.